### PR TITLE
[MRG] Transform the COBRE downloader to load individual files per subject

### PR DESCRIPTION
**Pull request by AlexandreAbraham**
In #915, @GaelVaroquaux stated that downloads lasting more than 1 hour were a problem. This PR address this problem by downloading individual files for COBRE dataset.

### CODE REVIEW

**Comment by KamalakerDadi** [comment]
What is the idea of this PR ? 

>This PR address this problem by downloading individual files for COBRE dataset.

But, if I want to download 146 subjects ? Anyway it will take 1 hour right ?

**Comment by AlexandreAbraham** [comment]
The goal is to be able to work with less than 146 subjects. I think that this is important for brainhack where the internet access may not be stable.

**Comment by AlexandreAbraham** [comment]
Travis failure is due to memory check.

File: nilearn/datasets/func.py [hunk 1]
```diff
@@ -1470,24 +1471,47 @@ def fetch_cobre(n_subjects=10, data_dir=None, url=None, verbose=1):
...
-    csv_array = np.recfromcsv(csv_file[0], names=names, skip_header=True)
-    # Get the ids of the datasets
-    ids = csv_array['subject_type']
-    max_subjects = len(ids)
+    names = ['id', 'sz', 'age', 'sex', 'fd']
+    csv_array = np.recfromcsv(csv_file, names=names, skip_header=True)
+    # Change dtype of id and condition column
+    csv_array = csv_array.astype(
+        [('id', '|U17'),
+         ('sz', 'bool'),
```
Comment (KamalakerDadi) [inline nilearn/datasets/func.py]: Any reason why you convert 'sz' ```int``` to ```bool``` ? and 'age' ```int``` to ```float``` ?

May be we should return as it is ? I assume csv file is prepared according to their needs.

Also, in the array dtype of 'sz' is not recognized '?'

File: nilearn/datasets/tests/test_func.py [hunk 2]
```diff
@@ -448,11 +448,37 @@ def test_fetch_cobre():
                104, 107, 111, 113, 114, 115, 116, 118, 119, 120, 121, 123,
                124, 125, 127, 128, 129, 130, 131, 134, 135, 136, 138, 139,
                140, 141, 144, 146, 147]
-    ids_sch = [('szxxx0040%03d' % i).encode() for i in ids_sc]
-    ids_cont = ids_cont = [('contxxx0040%03d' % i).encode() for i in ids_con]
-    subs = np.array(ids_sch + ids_cont, dtype='S17')
-    subs = subs.view(dtype=[('subject_type', 'S17')])
-    tst.mock_fetch_files.add_csv('cobre_model_group.csv', subs)
+    ids_sch = ['szxxx0040%03d' % i for i in ids_sc]
+    ids_cont = ids_cont = ['contxxx0040%03d' % i for i in ids_con]
```
Comment (KamalakerDadi) [inline nilearn/datasets/tests/test_func.py]: repeated ```ids_cont```

**Comment by KamalakerDadi** [comment]
LGTM

**Comment by KamalakerDadi** [comment]
Are you sure each individual link will not fail at some point ? I am not much aware of practical situations in these cases.

File: nilearn/datasets/func.py (hunk 1 above)
Comment (AlexandreAbraham) [inline nilearn/datasets/func.py]: I think it is good practice to put categorical variables (`sz` and `sex`) as integers and continuous variables (`age` and `fd`) as floats. The bool was an error. Age is not really converted, numpy loads everything as float anyway.

File: nilearn/datasets/tests/test_func.py (hunk 2 above)
Comment (AlexandreAbraham) [inline nilearn/datasets/tests/test_func.py]: Fixed.

**Comment by AlexandreAbraham** [comment]
> Are you sure each individual link will not fail at some point ? I am not much aware of practical situations in these cases.

We have no guarantee since we do not host the data. But `figshare` has an excellent API for managing datasets so I think that we should take advantage of it. The code to download COBRE (ie downloading the json and then each file individually) can be easily adapted to any other dataset hosted on figshare.

An option could be to rely on the big file if individual download fails but the nilearn code is not ready for it.

**Comment by bthirion** [comment]
LGTM too.

File: nilearn/datasets/func.py (hunk 1 above)
Comment (GaelVaroquaux) [inline nilearn/datasets/func.py]: > Any reason why you convert 'sz' int to bool ? and 'age' int to float ?

> May be we should return as it is ? I assume csv file is prepared according to
> their needs.

I am not sure: our goal is to make examples simple. To load CSVs, we
should really point users to pandas (we need to add notes about this here
and there in the documentation).

**Comment by GaelVaroquaux** [comment]
LGTM. Merging, thank you very much.

**Pull request merged**